### PR TITLE
Kopps cache

### PR DESCRIPTION
**Pull request by jhsware**
We now support a Redis cache for TKoppsCourseInfo to reduce calls to KOPPS.

- use Redis as cache for TKoppsCourseInfo
- use database 1 for cache and 0 for session
- clearly provide a legacy redis client for session (connect-redis)
- use NodeCache as fallback if Redis isn't available
- cache can easily be used for other data provided no key collisions

### CODE REVIEW

**Review by kaj — CHANGES_REQUESTED** [review]
Det mesta ser bra ut, jag har några undringar och några synpunkter ...

File: kpm/kpm-backend/src/api/commonCache.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// but only the fields we care about.
+// https://github.com/node-cache/node-cache
+// The standard ttl is given in seconds, I guess anything between 12
+// and 48 hours should be ok, maybe avoid purging stuff at the same
+// time every day by using 40 hours.
+const redisClient = hasRedis ? getRedisClient(REDIS_DB_NAMES.KOPPS) : null;
+const kopps_cache = hasRedis ? null : new NodeCache({ useClones: false });
+type TCacheable = object | string;
+
+export async function __CACHED_VALUE__<T extends TCacheable>(
```
[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts:19]: Det fula namnet är mitt sätt att få det att sticka ut i koden. Hade helst implementerat som decorator, men det betraktas fortf. experimentellt i JS/TS. Det är alltså vid användning jag vill att det ska sticka ut. Men känns det som en tagg i sidan så får jag ofc ge med mig :D

Gillar idén! Men då borde man nästan bryta ut det till ett eget paket som vi kan dela mellan alla API:er. Jag kikar på att göra ändringen i första steget. Då kan man även lägga in TTL redan där.

File: kpm/kpm-backend/src/api/commonCache.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  key: string,
+  ttlSecs: number,
+  fn: () => Promise<T | undefined>
+): Promise<T | undefined> {
+  const cachedValue = _getCachedValue<T>(key);
```
[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts:24]: Hjärnsläpp! Jag ska ta och skriva tester...

File: kpm/kpm-backend/src/api/commonCache.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * Code to CACHE Data
+ */
+// kopps_cache is a local cache of course_code -> kopps info object.
+// Note that we don't cache the entire (much larger) kopps response,
+// but only the fields we care about.
+// https://github.com/node-cache/node-cache
+// The standard ttl is given in seconds, I guess anything between 12
+// and 48 hours should be ok, maybe avoid purging stuff at the same
+// time every day by using 40 hours.
+const redisClient = hasRedis ? getRedisClient(REDIS_DB_NAMES.KOPPS) : null;
```
[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts:15]: Agree!

File: kpm/kpm-backend/src/redisClient.ts [hunk 4]
```diff
@@ -15,10 +15,29 @@ const useRedis = !!process.env.REDIS_HOST;
...
+export function getRedisClientForConnect(databaseName: REDIS_DB_NAMES) {
+  // connect-redis currently only supports v3 of Redis
+  // so it is compatible with ioredis. If this changes
+  // you can get the v4 Redis client by skipping legacy
+  // mode. https://github.com/tj/connect-redis/pull/337
+  return _getRedisClient(databaseName, true);
+}
 
+export function getRedisClient(databaseName: REDIS_DB_NAMES) {
+  // Return the v4 node-redis client
```
[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/redisClient.ts:32]: Will fix

[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/redisClient.ts:43]: Fixed in 24a57165f8

File: kpm/kpm-backend/src/api/commonCache.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts:8]: Fixed in d7dbf22095

File: kpm/kpm-backend/src/api/commonCache.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  key: string,
+  ttlSecs: number,
+  fn: () => Promise<T | undefined>
+): Promise<T | undefined> {
+  const cachedValue = _getCachedValue<T>(key);
+  if (cachedValue) return cachedValue;
+
+  const newValue = await fn();
+
+  if (newValue) _setCachedValue(key, newValue, ttlSecs);
```
[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts]: Fixed in cb791641dd

File: kpm/kpm-backend/src/api/commonCache.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts]: Fixed in cb791641dd

File: kpm/kpm-backend/src/api/commonCache.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  async cached(key: string): Promise<T | null | undefined> {
+    const cachedValue = await this._getCachedValue<T>(key);
+    if (cachedValue !== undefined) return cachedValue;
+
+    // Cache miss, calculate value, store and return
+    let newValue;
+    try {
+      newValue = await this._fn(key);
+    } catch (err) {
+      // TODO: Should we throw an error instead?
```
Comment (kaj) [inline kpm/kpm-backend/src/api/commonCache.ts:77]: If we want to provide commonCache as a package for others to use, we should probably just not catch the error, letting it propagate to the user.

If just for our own use case, I think we should log the error (at least as a warning, maybe as an error) in the catch.

File: kpm/kpm-backend/src/api/common.ts [hunk 7]
```diff
@@ -146,33 +143,26 @@ export async function getCourseInfo(
       )
       .then((r) => r.body);
```
Comment (kaj) [inline kpm/kpm-backend/src/api/common.ts:144]: Should we add a status code check here?  I don't think got throws an error on bad status code by default?

File: kpm/kpm-backend/src/api/commonCache.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+const redisClientInstances: Record<number, RedisClientType> = {};
+const nodeCacheInstances: Record<number, NodeCache> = {};
+
+/**
+ * Allow memoizing values with string as lookup key. Stores cached
+ * values in Redis if env var REDIS_HOST is set, otherwise uses
+ * in memory storage NodeCache.
+ */
+export class StringKeyCache<T extends TCacheable> {
```
Comment (kaj) [inline kpm/kpm-backend/src/api/commonCache.ts:23]: I know I asked for a method, but now that I see it, and it has only private properties, a constructor and one single public method, maybe it should be replaced with a function (the constructor) returning a function (the method)?

File: kpm/kpm-backend/src/api/commonCache.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    // Fall back to NodeCache in memory cache
+    if (this._redisClient === undefined) {
+      if (nodeCacheInstances[dbName] === undefined) {
+        nodeCacheInstances[dbName] = new NodeCache({ useClones: false });
+      }
+      this._nodeCache = nodeCacheInstances[dbName];
+    }
+  }
+
+  async cached(key: string): Promise<T | null | undefined> {
```
Comment (kaj) [inline kpm/kpm-backend/src/api/commonCache.ts:68]: Should we really have both `null` and `undefined` in the return type?  If so, we should at least define the intended semantic difference.  If it's just that redis and nodecache uses different return values for "not found", we should unify that.

**Review by kaj — COMMENTED** [review]
Some more questions here ...

File: kpm/kpm-backend/src/api/commonCache.ts (hunk 6 above)
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts]: Fixed in 27ae41626

File: kpm/kpm-backend/src/api/common.ts (hunk 7 above)
Comment (jhsware) [reply kpm/kpm-backend/src/api/common.ts]: Fixed in 1b2b63ca

File: kpm/kpm-backend/src/api/commonCache.ts (hunk 8 above)
Comment (jhsware) [reply kpm/kpm-backend/src/api/commonCache.ts]: Fixed in 4727279

**Review by kaj — APPROVED** [review]

**Pull request merged**